### PR TITLE
lesson: create type checking with PropTypes lesson

### PR DESCRIPTION
**Pull request by LeoKirasic**
## This PR

Creates a new PropTypes lesson for the current react revamp.

## Issue

Closes https://github.com/TheOdinProject/top-meta/issues/252

## Pull Request Requirements

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [x] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

File: react/react-ecosystem/type-checking-with-proptypes.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+### Introduction
+
+PropTypes is a way to type check the props that a React component receives. It helps to catch potential type errors during development, making it easier to spot and fix bugs. If you have used a linter in your previous React projects, there's a good chance it ended up yelling at you about certain props missing in prop validation, however if that isn't the case- well, lucky you!
```
Comment (01zulfi) [inline react/react-ecosystem/type-checking-with-proptypes.md:3]: We should start the introduction with a little explanation of what "type-checking"is. Afterwards we can introduce "PropTypes is a way ..."

File: react/react-ecosystem/type-checking-with-proptypes.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- Setting up PropTypes.
+- Using common PropTypes features.
+
+### Getting Started
+
+To start using PropTypes in our react projects, we first need to install the corresponding library. We can do that with ~~~npm~~~. In your react project run the following command:
```
Comment (01zulfi) [inline react/react-ecosystem/type-checking-with-proptypes.md:14]:
```suggestion
To start using PropTypes in our react projects, we first need to install the corresponding library. We can do that with `npm`. In your react project run the following command:
```

File: react/react-ecosystem/type-checking-with-proptypes.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+RenderName.propTypes = {
+  name: PropTypes.string,
+};
+
+export default RenderName;
+~~~
+
+In this example, the component RenderName expects to receive a prop called name which is a string. If this prop is not passed, or if it is not a string, a warning will be displayed.
```
Comment (01zulfi) [inline react/react-ecosystem/type-checking-with-proptypes.md:45]:
```suggestion
In this example, the component RenderName expects to receive a prop called `name` which is a string. If this prop is not passed, or if it is not a string, a warning will be displayed.
```

File: react/react-ecosystem/type-checking-with-proptypes.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+Now is also a good time to mention [TypeScript](typescriptlang.org) - a strongly typed language that builds on JavaScript. We don't cover it specifically, but it's worth learning about it if you'd like more safety while writing your code.
+
+If you go in the direction of learning TypeScript, our recommendation would be picking up a previous project and refactoring the components one by one to TypeScript. Learn by doing!
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Read through the [PropTypes documentation](https://reactjs.org/docs/typechecking-with-proptypes.html). It shows all of the types you can specify and some other useful things that can be done with it!
+2. Read [this article](https://blog.logrocket.com/validate-react-props-proptypes/) for a more in-depth look into the benefits and use cases of PropTypes.
```
Comment (01zulfi) [inline react/react-ecosystem/type-checking-with-proptypes.md:83]: Love this resource. Perhaps we can improve link text and provide a sneak peek of sorts?

```suggestion
2. You can even set up custom validators in PropTypes. Read [this article on using PropTypes on LogRocket](https://blog.logrocket.com/validate-react-props-proptypes/) for a more in-depth look into the benefits and use cases of PropTypes.
```

File: react/react-ecosystem/type-checking-with-proptypes.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+3. [This](https://stackoverflow.com/questions/41746028/proptypes-in-a-typescript-react-application) StackOverflow thread goes into the differences of TypeScript and PropTypes. Give it a read!
```
Comment (01zulfi) [inline react/react-ecosystem/type-checking-with-proptypes.md:84]: Improved link text:
```suggestion
3. This [StackOverflow thread goes into the differences of TypeScript and PropTypes](https://stackoverflow.com/questions/41746028/proptypes-in-a-typescript-react-application). Give it a read!
```

File: react/react-ecosystem/type-checking-with-proptypes.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+export default RenderName;
+~~~
+
+In this example, with the help of the defaultProps property we are defining a default value for the `name` prop. This way, if the `RenderName` component is called without passing in the `name` prop, it will default to "Zach". When you do pass in props, they will take precedence over the default props.
+
+### What about TypeScript?
+
+Now is also a good time to mention [TypeScript](typescriptlang.org) - a strongly typed language that builds on JavaScript. We don't cover it specifically, but it's worth learning about it if you'd like more safety while writing your code.
+
+If you go in the direction of learning TypeScript, our recommendation would be picking up a previous project and refactoring the components one by one to TypeScript. Learn by doing!
```
Comment (01zulfi) [inline react/react-ecosystem/type-checking-with-proptypes.md:76]: I'm slightly hesitant on providing a "how to learn TypeScript" recommendation as it _might_ hold us accountable for providing support (for TS) in our Discord chats. I'd like to hear your thoughts though

**Review by 01zulfi — CHANGES_REQUESTED** [review]
Great work :fire: :fire: :fire: 

I've left some comments below. Additionally, we should capitalize "react" anytime it's mentioned.

File: react/react-ecosystem/type-checking-with-proptypes.md (hunk 6 above)
Comment (LeoKirasic) [reply react/react-ecosystem/type-checking-with-proptypes.md:76]: @01zulfi I had the same doubts, however if we are bringing up TypeScript so learners are aware of it I feel uneasy at the very least not pointing them in the right direction. 

people often ask for advice on how to start with TS, and i always tell them the same thing. 

i would love if we did mention it this way. Perhaps adding a disclaimer how this is just a recommendation and that we don't explicitly cover or provide support for TS?

**Comment by LeoKirasic** [comment]
thanks for the suggestions, gonna work on incorporating those!

**Comment by LeoKirasic** [comment]
@01zulfi committed the changes. How do you like the short intro on type checking?

File: react/react-ecosystem/type-checking-with-proptypes.md (hunk 4 above)
Comment (LeoKirasic) [reply react/react-ecosystem/type-checking-with-proptypes.md]: could you elaborate more on the sneak peek you're referring to? I'm not sure that we have historically given sneak peeks in this section. Were you referring to the custom validators part or were you thinking something more along the lines of screenshots? 

sorry for the confusion on my part !

Comment (01zulfi) [reply react/react-ecosystem/type-checking-with-proptypes.md]: > Were you referring to the custom validators part

yup, that's definitely what I meant

File: react/react-ecosystem/type-checking-with-proptypes.md (hunk 6 above)
Comment (01zulfi) [reply react/react-ecosystem/type-checking-with-proptypes.md:78]: I see what you're saying.

A couple of suggestions to improve this section:

- We could rephrase the "We don't cover it specifically," to "We don't cover it in our curriculum yet,"

- At this stage of the curriculum, learning TS can be a lot of overhead when learners are already learning React. Perhaps we can incorporate this into our recommendation:

> Learning TypeScript can be a lot of overhead when you're already learning React. Although, once you're comfortable navigating the ins and outs of React, you'll have a better time learning and applying TypeScript to your projects. If you go in the direction of learning TypeScript, our recommendation would be picking up a previous project and refactoring the components one by one to TypeScript. Learn by doing!

Comment (LeoKirasic) [reply react/react-ecosystem/type-checking-with-proptypes.md]: @01zulfi done ! :rocket:

**Review by 01zulfi — APPROVED** [review]
:rocket: 🚀

**Pull request merged**